Large modular logic refactoring

Cyrilvallez opened the pull request:
# What does this PR do?

This PR largely rework the logic we use in the modular converter. It is (hopefully) clearer and maintainable. Instead of going in all directions, adding stuff, then deleting it if not needed, we now do the following:

- visit all the modular file (record imports/functions/classes/assignments nodes)
  - create function dependency mapping
- for each import coming from another model:
  - visit the corresponding file
  - create function dependency mapping
  - update mapping with function/assignment from the modular (updated/new functions)
  - create the class dependency graph based on merged dependencies
- update dependency graph of the modular with the functions and assignments imported from the other files
- for each class recorded in the modular:
  - if inherithing from class in another file:
    - replace call to super
    - find the dependencies after the node was replaced
    - follow (updated with modular defs) dependency mapping to add all nodes
  - else:
    - only add needed imported functions (and their dependencies)
- determine the needed imports and add them

Note that we now only visit each files once, instead of potentially revisiting them multiple times due to renaming or deleting nodes at the end.
cc @ArthurZucker for the logic design

Still yet to come if the design looks good:
- Unit-tests!!!

Cyrilvallez committed:
rework converter

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
cleaning

Cyrilvallez committed:
cleaning

Cyrilvallez committed:
finalize imports

Cyrilvallez committed:
imports

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
Better renaming to avoid visiting same file multiple times

Cyrilvallez committed:
start converting files

Cyrilvallez committed:
style

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_34487). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
-class ModularConverterTransformer(CSTTransformer):
-    METADATA_DEPENDENCIES = (ParentNodeProvider, ScopeProvider, PositionProvider)
+class ModularFileMapper(ModuleMapper):
+    """This is a Mapper for a modular file. It visits the whole file, recording dependency, then visits all model-specific
+    files that should be visited, and manages their mutual dependencies.
+    Calling the method `create_modules()` after visit will create all modules based on this modular file.
+    """
```
```suggestion
    """This is a Mapper to visit a modular file ( like `modular_llama.py`). It visits the whole file, recording dependency, then visits all imported
    files from the package (`transformers`) and manages their mutual dependencies.
    Calling the method `create_modules()` after visit will create all modules based on this modular file.
    """
```

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
-        2. Parse it into an AST Tree
-        3. Add this import to `self.transformers_imports` as visited to not parse it twice
+        """When visiting imports from model-specific files (i.e. `transformers.models.xxx`) we get the code, parse it,
+        and record it in `self.model_specific_modules`. The imported objects are recorded in `self.model_specific_imported_objects`.
```
not sure what record means in this context

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
+        and record it in `self.model_specific_modules`. The imported objects are recorded in `self.model_specific_imported_objects`.
         """
         import_statement = self.python_module.code_for_node(node.module)
+        if "auto.modeling_auto" in import_statement:
```
we need a general variable for this!

ArthurZucker replied:
(Like imports_to_skip_in_modular)

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -808,322 +1005,261 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
-        then we remove it from the new body to add it again in the correct order.
-        3. Replace the calls to `super().xxxx` merging parent code
+    def visit_SimpleStatementLine(self, node):
+        """If we visit an import statement not previously visited, record it. If we visit a top-level assignment,
```
top-level -> module scoped

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -808,322 +1005,261 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+                ):
+                    self.imports.append(node)
+            elif m.matches(node, simple_top_level_assign_structure):
+                assigned_variable = node.body[0].targets[0].target.value
+                # __all__ is treated differently and not added to general assignments
+                if assigned_variable != "__all__":
+                    self.assignments[assigned_variable] = node
+                else:
+                    assign_node = node.body[0]
+                    if isinstance(assign_node.value, cst.List):
+                        # Extract the elements from the list
+                        all_all_to_add = defaultdict(list)
+                        for element in assign_node.value.elements:
+                            if isinstance(element.value, cst.SimpleString):
+                                # Remove quotes and add the string to the elements list
+                                class_name = element.value.value
+                                file = find_file_type(element.value.evaluated_value)
+                                all_all_to_add[file] += [class_name]
+                        for file, new_alls in all_all_to_add.items():
+                            new_node = assign_node.with_changes(
+                                value=cst.List(
+                                    elements=[cst.Element(value=cst.SimpleString(value=k)) for k in new_alls]
+                                )
+                            )
+                            self.all_all_to_add[file] = node.with_changes(body=[new_node])
```
let's maybe create a separate function for whatever does not need self in this? (would make it more readable)

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -808,322 +1005,261 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
-        return node
+    def leave_Module(self, node):
+        """When we leave the modular file, we do the following in order:
+        - compute recursive function dependencies
+        - for each model-specific file found in the imports, rename it with the new model name, visit it, and update
+        its dependency graph with the new function and assignment definitions found in the modular
+        - update the modular dependency graph with the imported functions and assignments (found when visiting the matching files)
```
```suggestion
        1. compute the nested function dependencies
        2. for each model-specific file found in the imports, rename it with the new model name, visit it, and update
        its dependency graph with the new function and assignment definitions found in the modular
        3. update the modular dependency graph with the imported functions and assignments (found when visiting the matching files)
```
would be nice if these are reflected with # 1. , # 2. etc as comments

ArthurZucker replied:
>  for each model-specific file found in the imports, rename it with the new model name: for this we parse the file, and use a FileModelMapper, to visit it and upsate the model name, dependency graph etc

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -808,322 +1005,261 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+                self.added_objects_file_mapping[object_name] = file
+                self.assignments[object_name] = visited_module.assignments[object_name]
+
+    def compute_relative_order(self, missing_dependencies: set) -> dict[str, int]:
```
does this need to be in this visiter? probably not no?

Cyrilvallez replied:
As the goal anf functionality is the same but the logic is different depending on the type of `Mapper`, I think it makes sense to keep it in the visitors. (It uses a lot of `self` internal variables as well)

ArthurZucker commented on utils/modular_model_converter.py:
```diff
@@ -808,322 +1005,261 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+
+        return nodes_to_add, file_type
+
+    def create_modules(self) -> dict[str, cst.Module]:
```
same here, if we create the modules post `leave_Module` I don't think this should be in this class no?

Cyrilvallez replied:
Yes I was not sure about this choice, I kept it in the class because it was using a lot of internal variables. But as we move more and more stuff outside, let's keep a functional approach.

ArthurZucker reviewed (commented):
Okay! Very nice idea to have a general ModelMapper, and have one for Modular files and one for "other" files. 

Let's keep splitting the functionalities, I think creating the module should go outside the visiter, don't necessarily need a class ! 

Looks good in general! Great work 🔥

Cyrilvallez committed:
address most comments

Cyrilvallez committed:
style

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -524,11 +376,381 @@ def leave_Return(self, original_node: cst.Return, updated_node: cst.Return) -> c
+        self.classes[node.name.value] = node
+
+    def visit_Call(self, node: cst.Call):
+        """This is used to create a mapping from top-level functions to functions called inside them.
```
Yes, indeed in all generality it should be a `Name`. I changed and made sure to update at the end so that we don't keep all the garbage Names we don't want (i.e. built-ins, imports etc)

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
+                    unused_imports.add(name)
+                import_ref_count[name] = ref_count
+
+    # Note that dicts implicitly keep the order of insertion
```
Starting in 3.7, which is end of life now, so should be fine. Anyway, the order is actually not super important here as of now, as the imports are not yet sorted with proper logic (simply relies on the order it is seen in the files, but since we merge imports from different files, the order may end-up being slightly surprising)

Cyrilvallez replied:
Just removed the dicts to use a list anyway!

yonigozlan commented:
Hey, so nice to see modular getting better and better! Just a heads up that I have this relatively short PR out on modular as well https://github.com/huggingface/transformers/pull/34477 , which adds some functionalities needed for the ColPali PR, so I was wondering if it will be compatible with this refactoring!
Also made a comment there on adding the modular examples to the check_modular_conversion script to traced when some issues are introduced, I think it could be useful to have here also :)

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
+def append_new_import_node(
+    node: cst.CSTNode, unused_imports: set[str], imports_to_keep: dict[str, cst.CSTNode], current_idx: int
+):
+    """Insert the new `node` to the dict of `imports_to_keep` in-place, if it is not part of the `unused_imports`.
+    This function takes cares of aggregating similar ImportFrom, i.e. if we ever saw a statement such as
+    `from typing import Any`, and later another one `from typing import List`, we will aggregate as
+    `from typing import Any, List` in a single statement.
```
At some point I was still generating files with duplicated lines of the same imports 
```python
from typing import Optional, Tuple, Union
...
from typing import List
```
which is why I added this quick fix. It does not seem to be needed anymore with latest development, so I removed it for simplicity.
I don't know why it was happening then 😅

Cyrilvallez committed:
remove unused stuff in get_needed_imports

Cyrilvallez committed:
style

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -635,170 +868,134 @@ def replace_call_to_super(
+    elif m.matches(node, m.Attribute()):
+        # Recursively get the full name for attributes
+        return f"{get_module_name(node.value)}.{node.attr.value}"
+    return ""
```
This function was removed

Cyrilvallez committed:
move class dependency functions outside class

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -524,11 +376,381 @@ def leave_Return(self, original_node: cst.Return, updated_node: cst.Return) -> c
+        wrapper = MetadataWrapper(temp_module)
+        visitor = cls(updated_node.name.value, set(mapper.global_nodes.keys()))
+        wrapper.visit(visitor)
+        return mapper.augment_dependencies_with_functions(visitor.dependencies)
```
I moved them outside the class and slightly renamed them.

Cyrilvallez committed:
Move main functions outside class

Cyrilvallez committed:
style

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
rename func

Cyrilvallez committed:
add augmented dependencies

Cyrilvallez committed:
Update modular_model_converter.py

Cyrilvallez committed:
Add types_to_file_type + tweak annotation handling

Cyrilvallez commented:
Hey @yonigozlan thanks for the heads-up! I added the new `TYPE_TO_FILE_TYPE` and slightly tweaked how I was handling `Annotations`, your `modular_new_kwargs_model.py` now behaves correctly also in this PR.
If you have any issue let me know.

ArthurZucker reviewed (approved):
Very very nice! 
The only things missing now: 
- unittests /  small examples of the capabilities
- update the documentation a little bit to further explain how we do this but mostly 
To be honest we can merge this will unblock me as well 🤗

Cyrilvallez committed:
Allow assignment dependency mapping + fix regex

Cyrilvallez committed:
style + update modular examples

Cyrilvallez committed:
fix modular_roberta example (wrong redefinition of __init__)

Cyrilvallez commented:
Hey @yonigozlan! Thanks for the feedback! I added **assignment dependency tracking** to correctly follow everything (I previously thought it would never be needed, but it turns out it is! And good to have it in general anyway).  
Everything is now behaving nicely, both in the actual models definitions and examples (the examples were broken already btw). Could you double-check just in case? This highlights that we really need those **tests** soon!  
I'll let @ArthurZucker decide if we want to add the examples in the `check_conversion`, but I added them to the `--files_to_parse all` anyway.

Cyrilvallez committed:
slightly correct order in which dependencies will appear

Cyrilvallez committed:
style

Cyrilvallez committed:
review comments

Cyrilvallez committed:
Performance + better handling of dependencies when they are imported

Cyrilvallez committed:
style

Cyrilvallez committed:
Add advanced new classes capabilities

Cyrilvallez committed:
style

Cyrilvallez committed:
add forgotten check

Cyrilvallez commented:
- Added support for a new important use-case (see https://github.com/huggingface/transformers/pull/33770): **correct dispatch of fully new class**. Consider the following `modular.py`:
```
    from ..llama.modeling_llama import LlamaModel

    class NewNameTextConfig(PretrainedConfig):
        ...

    class NewNameConfig(PretrainedConfig):
        ...

    class NewNameModel(LlamaModel):
        config = NewNameConfig()
        text_config = NewNameTextConfig()
        ...
```
we previously had no way of correctly dispatching `NewNameTextConfig` to `configuration_newname.py`, without importing it in `modeling_newname.py` as well as part of the dependencies (because `modeling_llama.py` only tell us that `NewNameConfig` will be imported, but nothing about `TextConfig`).
This is now solved, every fully new class (without exact name match e.g. `LlamaConfig` <-> `NewNameConfig`) which does not belong to the correct file type will not be added, and an import will be created instead.

Cyrilvallez committed:
Update modeling_llava_next_video.py

Cyrilvallez commented on utils/modular_model_converter.py:
```diff
@@ -632,173 +894,144 @@ def replace_call_to_super(
     "Processor": "processing",
     "ImageProcessor": "image_processing",
     "FeatureExtractor": "feature_extractor",
+    "ProcessorKwargs": "processing",
+    "ImagesKwargs": "processing",
+    "TextKwargs": "processing",
```
Are we expecting clashes in the names here? Because infering based on the class using them is not so straightforward since different classes in different file types may use them (e.g. type hints)

Cyrilvallez replied:
For what I can see, all of them are only used in "processing" files as of now at least

Cyrilvallez committed:
Add prority list ordering in check_conversion as well

Cyrilvallez committed:
Update check_modular_conversion.py

Cyrilvallez committed:
Update configuration_gemma.py

ArthurZucker commented:
🥳
